zebra-consensus/src/transaction.rs

```diff
@@ -522,31 +503,66 @@ where
                 //.map_err(|e| BoxError::from(Box::new(e)))?;
             }
 
-            let _rsp = redjubjub_verifier
-                .ready_and()
-                .await?
-                .call((bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into())
-                .boxed();
-
             // TODO: stop ignoring binding signature errors - #1939
-            // async_checks.push(rsp);
+            // async_checks.push(
+            //     primitives::redjubjub::VERIFIER
+            //         .clone()
+            //         .oneshot((bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into()),
+            // );
         }
 
         Ok(async_checks)
     }
+}
+
+/// A set of unordered asynchronous checks that should succeed.
+///
+/// A wrapper around [`FuturesUnordered`] with some auxiliary methods.
+struct AsyncChecks(FuturesUnordered<Pin<Box<dyn Future<Output = Result<(), BoxError>> + Send>>>);
+
+impl AsyncChecks {
+    /// Create an empty set of unordered asynchronous checks.
+    pub fn new() -> Self {
+        AsyncChecks(FuturesUnordered::new())
+    }
+
+    /// Push a check into the set.
+    pub fn push(&mut self, check: impl Future<Output = Result<(), BoxError>> + Send + 'static) {
+        self.0.push(check.boxed());
+    }
 
-    /// Await a set of checks that should all succeed.
+    /// Push a set of checks into the set.
+    ///
+    /// This method can be daisy-chained.
+    pub fn and(mut self, checks: AsyncChecks) -> Self {
+        self.0.extend(checks.0);
+        self
+    }
+
+    /// Wait until all checks in the set finish.
     ///
     /// If any of the checks fail, this method immediately returns the error and cancels all other
     /// checks by dropping them.
-    async fn wait_for_checks(mut checks: AsyncChecks) -> Result<(), TransactionError> {
+    async fn check(mut self) -> Result<(), BoxError> {
         // Wait for all asynchronous checks to complete
         // successfully, or fail verification if they error.
-        while let Some(check) = checks.next().await {
-            tracing::trace!(?check, remaining = checks.len());
+        while let Some(check) = self.0.next().await {
+            tracing::trace!(?check, remaining = self.0.len());
             check?;
         }
 
         Ok(())
     }
 }
+
+impl<F> FromIterator<F> for AsyncChecks
+where
+    F: Future<Output = Result<(), BoxError>> + Send + 'static,
+{
+    fn from_iter<I>(iterator: I) -> Self
+    where
+        I: IntoIterator<Item = F>,
+    {
+        AsyncChecks(iterator.into_iter().map(FutureExt::boxed).collect())
+    }
+}
```

Comment:
Style nit: I personally order types in a module so that a type is declared before it is used, so that would have `AsyncChecks` put at the top of the module where the original type alias was. But it has grown, so maybe it should be in its own module (`transaction.rs` is doing a lot anyway). Thoughts? I don't know if Rust has a style guide in this regard.

Reply:
I agree that this file is getting larger, and we should move some of the code into separate submodules.

But this PR is a low-priority refactor, so I'd rather open a cleanup ticket, merge this PR, and finish up #1981.

Reply:
Tracking in https://github.com/ZcashFoundation/zebra/issues/2436